extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/RunWorkOrder.java

```diff
@@ -215,23 +172,11 @@ public void startAsync()
       throw new ISE("Cannot start from state[%s]", state);
     }
 
-    final StageDefinition stageDef = workOrder.getStageDefinition();
-
     try {
       exec.registerCancellationId(cancellationId);
-      makeInputSliceReader();
-      makeWorkOutputChannelFactory();
-      makeShuffleOutputChannelFactory();
-      makeAndRunWorkProcessors();
-
-      if (stageDef.doesShuffle()) {
-        makeAndRunShuffleProcessors();
-      } else {
-        // No shuffling: work output _is_ stage output. Retain read-only versions to reduce memory footprint.
-        stageOutputChannelsFuture =
-            Futures.immediateFuture(workResultAndOutputChannels.getOutputChannels().readOnly());
-      }
-
+      initInputSliceReader();
```

Comment:
I think the `ExecutionContext` is created to late; for example this `initInputSliceReader` method is called twice - here and from when that context is created.

I feel like creating it earlier could enable to clean up some of this logic